dotnet/src/Connectors/Connectors.Onnx/BertOnnxTextEmbeddingGenerationService.cs

```diff
@@ -0,0 +1,284 @@
+        Stream onnxModelStream,
+        Stream vocabStream,
+        BertOnnxOptions? options = null)
+    {
+        Task<BertOnnxTextEmbeddingGenerationService> t = CreateAsync(onnxModelStream, vocabStream, options, async: false, cancellationToken: default);
+        Debug.Assert(t.IsCompleted);
+        return t.GetAwaiter().GetResult();
+    }
+
+    /// <summary>Creates a new instance of the <see cref="BertOnnxTextEmbeddingGenerationService"/> class.</summary>
+    /// <param name="onnxModelPath">The path to the ONNX model file.</param>
+    /// <param name="vocabPath">The path to the vocab file.</param>
+    /// <param name="options">Options for the configuration of the model and service.</param>
+    /// <param name="cancellationToken">The <see cref="CancellationToken"/> to monitor for cancellation requests. The default is <see cref="CancellationToken.None"/>.</param>
+    public static Task<BertOnnxTextEmbeddingGenerationService> CreateAsync(
+        string onnxModelPath,
+        string vocabPath,
+        BertOnnxOptions? options = null,
+        CancellationToken cancellationToken = default) =>
+        CreateAsync(onnxModelPath, vocabPath, options, async: true, cancellationToken: default);
+
+    /// <summary>Creates a new instance of the <see cref="BertOnnxTextEmbeddingGenerationService"/> class.</summary>
+    /// <param name="onnxModelStream">Stream containing the ONNX model.</param>
+    /// <param name="vocabStream">Stream containing the vocab file.</param>
+    /// <param name="options">Options for the configuration of the model and service.</param>
+    /// <param name="cancellationToken">The <see cref="CancellationToken"/> to monitor for cancellation requests. The default is <see cref="CancellationToken.None"/>.</param>
+    public static Task<BertOnnxTextEmbeddingGenerationService> CreateAsync(
+        Stream onnxModelStream,
+        Stream vocabStream,
+        BertOnnxOptions? options = null,
+        CancellationToken cancellationToken = default) =>
+        CreateAsync(onnxModelStream, vocabStream, options, async: true, cancellationToken: default);
+
+    private static async Task<BertOnnxTextEmbeddingGenerationService> CreateAsync(
+        string onnxModelPath,
+        string vocabPath,
+        BertOnnxOptions? options,
+        bool async,
+        CancellationToken cancellationToken)
+    {
+        Verify.NotNullOrWhiteSpace(onnxModelPath);
+        Verify.NotNullOrWhiteSpace(vocabPath);
+
+        using Stream onnxModelStream = new FileStream(onnxModelPath, FileMode.Open, FileAccess.Read, FileShare.Read, 1, async);
+        using Stream vocabStream = new FileStream(vocabPath, FileMode.Open, FileAccess.Read, FileShare.Read, 1, async);
+
+        return await CreateAsync(onnxModelStream, vocabStream, options, async, cancellationToken).ConfigureAwait(false);
+    }
+
+    private static async Task<BertOnnxTextEmbeddingGenerationService> CreateAsync(
```

Comment:
Can we delay this initialization on the service execution (when calling `GenerateEmbeddingsAsync`? So we avoid executing Async as sync on the service + kernel initialization?

Reply:
We _could_, but:
a) That would complicate the implementation, which would need to handle the concurrency issues associated with ensuring that it was only initialized once.
b) Any errors associated with configuration wouldn't be discovered until use, e.g. if the path to the model didn't exist.
c) That complicates ownership when constructed from Streams. Currently I have it set up where the caller passing in a Stream retains ownership of the stream and can dispose of it after the Add/Create calls return. But if the initialization were delayed until possible first use, ownership would need to be transferred.

> So we avoid executing Async as sync on the service + kernel initialization?

The synchronous code paths use synchronous I/O.